Core/src/GapicRequestWrapper.php

```diff
@@ -0,0 +1,248 @@
+<?php
+/**
+ * Copyright 2023 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Core;
+
+use Google\Cloud\Core\Exception;
+use Google\ApiCore\ApiException;
+use Google\ApiCore\OperationResponse;
+use Google\ApiCore\PagedListResponse;
+use Google\ApiCore\Serializer;
+use Google\ApiCore\ServerStream;
+use Google\Protobuf\Internal\Message;
+use Google\Rpc\BadRequest;
+use Google\Rpc\Code;
+use Google\Rpc\RetryInfo;
+use LogicException;
+
+/**
+ * The GapicRequestWrapper is responsible for delivering REST/gRPC requests.
+ */
+class GapicRequestWrapper
+{
+    use RequestWrapperTrait;
+    /**
+     * @var Serializer A serializer used to encode responses.
+     */
+    private $serializer;
+
+    /**
+     * @var array gRPC specific configuration options passed off to the ApiCore
+     * library.
+     */
+    private $requestOptions;
+
+    /**
+     * @var array Map of error metadata types to RPC wrappers.
+     */
+    private $metadataTypes = [
+        'google.rpc.retryinfo-bin' => RetryInfo::class,
+        'google.rpc.badrequest-bin' => BadRequest::class
+    ];
+
+    /**
+     * @param array $config [optional] {
+     *     Configuration options. Please see
+     *     {@see Google\Cloud\Core\RequestWrapperTrait::setCommonDefaults()} for
+     *     the other available options.
+     *
+     *     @type Serializer $serializer A serializer used to encode responses.
+     *     @type array $requestOptions Configuration options passed
+     *           off to the ApiCore library.
+     * }
+     */
+    public function __construct(array $config = [])
+    {
+        // @TODO: Do we need this?
+        $this->setCommonDefaults($config);
+
+        // convert retry params to general params here
+        // @TODO: Do we need this?
+        $config = $this->convertConfigParams($config);
```

Comment:
We probably don't need this, see https://github.com/saranshdhingra/google-cloud-php/pull/5/files#r1410361984